### ISSUE TITLE
Dropping connection from spectator server mid-play results in truncated online replay

### ISSUE THREAD

**Issue by rvnexe** [ppy/osu-server-spectator#244]
Hi. I've downloaded a replay file from my profile to send it for someone, but it displays only a few seconds 'till the end.
This zip file contains the original replay file, which game client gave me. And the replay file which I downloaded from osu.ppy.sh website.
[replayfiles.zip](https://github.com/user-attachments/files/17057704/replayfiles.zip)

**Comment by peppy** [comment]
The replay retrieved from web contains frames from 658 frames from 141,060 to 152,154 ms, towards the latter end of the beatmap.

This implies the saved replay was truncated. Likely a failed connection to `osu-server-spectator`.

I'm not sure how we should be handling this yet. Either the whole play should be invalidated, or maybe the client needs to send catch-up data. (low priority cc/ @bdach for visibility on this).

**Comment by smoogipoo** [comment]
We could probably send the whole replay again after completion for validation?

**Comment by bdach** [comment]
I swear I've seen this reported somewhere already and even recall reproducing a connection drop-out doing this but I can't find it now. Maybe a github account deletion sent it to the shadow realm.

**Comment by peppy** [comment]
> We could probably send the whole replay again after completion for validation?

Maybe yeah. It's not a high prio item since this should happen quite rarely, but requires some discussion on validation concerns (ie. whether we want to use real-time packets as score validation piece).

**Comment by bdach** [comment]
Rough initial plan of attack detailed below.

## Happy path
```mermaid
sequenceDiagram
	client->>server: BeginPlaySession(scoreToken: 1234, state)
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 1, ... })
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 2, ... })
	server-->>client: ACK
	
	Note over client,server: ...
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 100, ... })
	server-->>client: ACK
	
	client->>server: EndPlaySession(state: { bundlesSent: 100, ... })
	server-->>client: ACK
	
	server->>client: CompleteReplay(scoreToken: 1234, bundles: [])
	activate client
	client-->>server: ACK
	activate server
	Note right of client: Client purges all bundles<br/>related to score token id=1234 from local buffers.
	deactivate client
	
	server->>S3: Upload replay
	S3-->>server: 200 OK
	Note left of server: Server purges all bundles<br/>related to score token id=1234 from local buffers.
	deactivate server
```
## Client goes away and returns later
```mermaid
sequenceDiagram
	client->>server: BeginPlaySession(scoreToken: 1234, state)
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 1, ... })
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 2, ... })
	server-->>client: ACK
	
	Note over client,server: After a while, connection is lost.
	
	client-xserver: SendFrameData(bundle: { sequenceNumber: 40, ... })
	
	client-xserver: SendFrameData(bundle: { sequenceNumber: 41, ... })
	
	Note over client,server: After a while, connection is re-established.
	
	client->>server: BeginPlaySession(scoreToken: 1234, state)
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 76, ... })
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 77, ... })
	server-->>client: ACK
	
	Note over client,server: This continues until gameplay completes.
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 100, ... })
	server-->>client: ACK
	
	client->>server: EndPlaySession(state: { bundlesSent: 100, ... })
	server-->>client: ACK
	
	server->>client: CompleteReplay(scoreToken: 1234, bundles: [40, 41, ..., 75])
	activate client
	client-->>server: [...bundles]
	activate server
	Note right of client: Client purges all bundles<br/>related to score token id=1234 from local buffers.
	deactivate client
	
	server->>S3: Upload replay
	S3-->>server: 200 OK
	Note left of server: Server purges all bundles<br/>related to score token id=1234 from local buffers.
	deactivate server
```

The above should also cover the client coming online midway through a play.

The minimum requirement here for a replay to be present is for at least *one* `BeginPlaySession()` call to go through while the play is in progress.
## Client goes away and returns later, but cannot provide missing frames
```mermaid
sequenceDiagram
	client->>server: BeginPlaySession(scoreToken: 1234, state)
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 1, ... })
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 2, ... })
	server-->>client: ACK
	
	Note over client,server: After a while, connection is lost.
	
	client-xserver: SendFrameData(bundle: { sequenceNumber: 40, ... })
	
	client-xserver: SendFrameData(bundle: { sequenceNumber: 41, ... })
	
	Note over client,server: After a while, connection is re-established.
	
	client->>server: BeginPlaySession(scoreToken: 1234, state)
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 76, ... })
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 77, ... })
	server-->>client: ACK
	
	Note over client,server: This continues until gameplay completes.
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 100, ... })
	server-->>client: ACK
	
	client->>server: EndPlaySession(state: { bundlesSent: 100, ... })
	server-->>client: ACK
	
	server->>client: CompleteReplay(scoreToken: 1234, bundles: [40, 41, ..., 75])
	activate client
	Note right of client: For whatever reason frames cannot be returned (e.g. old client).
	client-->>server: []
	activate server
	deactivate client
	
	server->>S3: Upload partial replay
	S3-->>server: 200 OK
	Note left of server: Server purges all bundles<br/>related to score token id=1234 from local buffers.
	deactivate server
```
## Client goes away and never returns
```mermaid
sequenceDiagram
	client->>server: BeginPlaySession(scoreToken: 1234, state)
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 1, ... })
	server-->>client: ACK
	
	client->>server: SendFrameData(bundle: { sequenceNumber: 2, ... })
	server-->>client: ACK
	
	Note over client,server: After a while, connection is lost.<br/>The client does not return for $TIMEOUT,<br/>where $TIMEOUT is relative to the time<br/>of last frame bundle received.<br/>This works because the client is expected<br/>to send bundles at least every 200ms.
	
	server->>S3: Upload partial replay
	activate server
	S3-->>server: 200 OK
	Note left of server: Server purges all bundles<br/>related to score token id=1234 from local buffers.
	deactivate server
```

## Scenarios where data loss occurs
- Not a single `BeginPlaySession()` call succeeded.
	- Retrying `BeginPlaySession()` is basically infeasible without preventing the user from playing again in the future until the retry succeeds. See also: https://github.com/ppy/osu/issues/24609.
- Not a single `EndPlaySession()` call succeeded.
	-  See above.
- The client has somehow irretrievably lost the frames on its side or fails to respond with the missing frames. (Two main relevant scenarios are: old client that doesn't buffer anything, and hard crash death.)
	- This is only limited to the frames that the server didn't receive.

**Comment by bdach** [comment]
I am unassigning from this, as after users from two major countries started recently mass-reporting complete and total disruption of communication with spectator server for no discernible reason my motivation to work on these fixes is zero.

**Comment by rvnexe** [comment]
Some problems are obviously country-specific; for instance, Iran's internet connection has some strict blockades on certain IP/domains, one of them is osu! which as far I've researched, it's because of some explicit content being hosted on there (backgrounds, avatars, etc. which idk it's our government's logic). Players here as faw as I now, use a wide range of VPN protocols (mostly v2ray-related) which of course can cause spec-server problems.

But it's so funny that only osu.ppy.sh subdomain is blocked by likely a SNI filtering (because direct IP pinging works.) so, for now, putting on some backup domains may be so helpful for people located here; because everything is operational but auth, replay submission, spectator, etc. And it may also help people around the globe who are messing with similar issues.

**Comment by rvnexe** [comment]
It might sound sus but I can host a very simple, small nodejs mirror located in Iranian datacenters to act as a proxy between players here and osu! servers to avoid similar problems in the future.

(You can also make your own; here a webservice provider called "ArvanCloud" hosts node.js snippets on Edge computing system (where codes run on the closest pop-site to user) with absolute $0 cost and no limitation in rate & count; also can bind subdomains with DNS record. Could be interesting to try; Their other services are so cheap; may one day, serve middle-asia region as middleware to osu! gameservers whatever)
